text/3466-manuallydrop-deref-move.md

```diff
@@ -0,0 +1,159 @@
+- Feature Name: `manuallydrop_deref_move`
+- Start Date: 2023-07-30
+- RFC PR: [rust-lang/rfcs#3466](https://github.com/rust-lang/rfcs/pull/3466)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+Extend the special-case move-out-of-deref behavior of `Box<T>` to `ManuallyDrop<T>`. Additionally, allow partial moves out of a `T` stored inside `ManuallyDrop<T>` even when there is a `Drop` impl for `T`.
+
+# Motivation
+
+Often, instead of dropping a struct, one wants to move out of one of its fields. However, this is impossible to do in safe code when the struct implements  `Drop`, requiring use of `unsafe` APIs like `std::ptr::read`, or runtime-expensive workarounds like wrapping the field in an `Option` and using `take()`.
+
+```rust
+pub struct Foo(String);
+
+impl Drop for Foo {
+    fn drop(&mut self) { /* ... */ }
+}
+
+pub struct Bar(String);
+
+impl Foo {
+    pub fn into_bar(self) -> Bar {
+        let m = core::mem::ManuallyDrop::new(foo);
+        Bar(unsafe { core::ptr::read(&m.0) }) // Need to use `unsafe`
+    }
+}
+```
+
+I ran into this limitation while working on the [`async-lock` library](https://github.com/smol-rs/async-lock/blob/8045684f996b15b3dd9bfd621cfc3864d3760923/src/rwlock.rs#L879-L883). It has also been discussed elsewhere, including [this blog post from 2018](https://phaazon.net/blog/rust-no-drop) (which recommends `mem::uninitialized` as a workaround!) and [this pre-RFC on Internals](https://internals.rust-lang.org/t/pre-rfc-destructuring-values-that-impl-drop/10450).
+
+# Explanation
+
+In today's Rust, `Box<T>` has the unique capability that it is possible to move out of a dereference of it.
+
+```rust
+let b: Box<String> = Box::new("hello world".to_owned());
+let s: String = *b; // `b`'s backing allocation is dropped here
+```
+
+Partial moves are also permitted.
+
+```rust
+let s: String;
+{
+    let b: Box<(String, String)> = Box::new(("hello".to_owned(), "world".to_owned()));
+    s = b.1;
+    // `b.0` and `b`'s backing allocation dropped here
+}
+```
+
+This RFC extends this capability to `ManuallyDrop`.
+
+```rust
+use core::mem::ManuallyDrop;
+
+let m: ManuallyDrop<(String, String)> = ManuallyDrop::new(("hello".to_owned(), "world".to_owned()));
+let _: String = m.1; // `m.1` dropped here
```

Comment:
```suggestion
drop(m.1); // `m.1` moved out of here
```
just `let _ =` never moves in the first place

Reply:
Thanks, fixed